### DISCUSSION TITLE
Welcome to asammdf Discussions!

### DISCUSSION THREAD

**Discussion by danielhrisca** [danielhrisca/asammdf#451] [General]
A place to ask questions and share knowledge

**Comment by RuwenSpitzley** [comment]
Hi everyone, 

batch processing many files I just want to extract single signals from huge 4.1 MDF4 files (sorted, I guess).
From my understanding MDF4() reads the meta-data to memory and later on the signal-samples data as required. correct?

MDF4() takes about 7 seconds in average and reading the single signal takes nothing at all.  
Is there a way to speed up? If not, do you think it is possible theoretically or does MDF-spec require full reading of all meta-data?  

thanks for your time!

**Reply by danielhrisca** [reply]
Hello @RuwenSpitzley 

> From my understanding MDF4() reads the meta-data to memory and later on the signal-samples data as required. correct?

Yes that is correct, the meta-data is loaded and the mininal info about the raw data blocks. 

To slow file load could caused by sorting the file or processing CAN bus logging. Can you check if you hit the _sort or _process_can_logging methods?

**Reply by RuwenSpitzley** [reply]
Hello @danielhrisca 

WOW! so quick reply!  

Well, I hit both, _sort and _process_can_logging, I quickly confirmed with debug-points 
After calling MDF4() it takes about 10s to enter debug-point at beginning of "def _sort(self): ..."
From there it takes about 0.5s  until the debug-point at "def _process_can_logging(self):" is reached. 
From there another 0.5s to come back to my test-script next statement (after MDF4-call )
My current file is 314 MB,  channel_db returns 16804 channel!   
I tried a bit more and the "while dg_addr:"  (mdf_v4.py -> row 469f)  consumes the effort.  11000 loops... 

Reasonable analysis? 
I hoped that it is possible to point to a individual signal-name and get it directly from the file. But as I now understand(actually guess) the channel-groups must be fully analyzed before a single signal can be extracted. correct?

**Reply by RuwenSpitzley** [reply]
Me again!
just realized that some logging-feature was switched on, when I first measure the code-speed (*shame on me*, 've been a long year)
running the test from terminal is around  4 times quicker. So loading the 314 MB test file takes 2.5s.  That*s far better then 10 to 15s... juhu! 
I still wonder if there is a way to speed up  (< 0.1s I'd like). E.g by putting above mentioned expensive loops in several threads? 

enjoy the night!

**Reply by danielhrisca** [reply]
> I still wonder if there is a way to speed up (< 0.1s I'd like). E.g by putting above mentioned expensive loops in several threads?

I'm afraid threading will not bring too much in this case, and it would make things a lot more complicated.

I would like to do some profiling using your file if it possible for you to share it. You can scramble it first to protect sensitive information (see [here](https://asammdf.readthedocs.io/en/latest/api.html#asammdf.mdf.MDF.scramble))

**Reply by danielhrisca** [reply]
> I still wonder if there is a way to speed up (< 0.1s I'd like). E.g by putting above mentioned expensive loops in several threads?

I have a try multi multithreaded parsing of data groups but there was no gain and the code got more complicated so I don't see a benefit for this

**Reply by danielhrisca** [reply]
> I hoped that it is possible to point to a individual signal-name and get it directly from the file. But as I now understand(actually guess) the channel-groups must be fully analyzed before a single signal can be extracted. correct?

@RuwenSpitzley 
I've managed to implement a channel loading filter that can be used. It does improve the loading time dramatically

```python
from asammdf import MDF

mdf = MDF('test.mf4', channels=["S1", "S2", "S3"])
```

You will need to install the development branch code to try it out

``pip install -I --no-deps https://github.com/danielhrisca/asammdf/archive/development.zip`` 
or
``pip install -I --no-deps git+https://github.com/danielhrisca/asammdf@development``

**Reply by RuwenSpitzley** [reply]
@danielhrisca 
wow. I also played around with seperate process/core and the result was not really satisfying yet; no fun in debugging... 
For sure, your approach will be the better solution. I'll try tonight with my messy and huge data and giver you feedback at once!

**Reply by danielhrisca** [reply]
@RuwenSpitzley did you try version 6.1.2?

**Comment by danielhrisca** [comment]
General tip: always use the MDF class. It provides several methods on top fo the ones provided by the MDF3 or MDF4 classes (depending on the file version).

**Comment by NicolaSabino** [comment]
Hi there

I'm using this library to process a huge file (5GB) composed of video frames and containing only one group and 12 indexes. I'm interested only in the 12th index.

The problem is that if I try to get these data with` file.get(group=0, index=12, raw=True)` because it is trying to load a whole Signal in RAM. 

**Is there a way to yield that specific Signal (group and index) sample by sample?**

**Reply by danielhrisca** [reply]
I've created a new discussion thread

**Comment by mini2WP** [comment]
I have been for some time trying to use the filter function in order to reduce a MF4 file (recorded with CANape 14 SP5). It works great until I try to extract the multimedia (non-embedded video) signal.

The error returned by v.6.2.0 is:
```
    File "C:\Python39\lib\site-packages\asammdf\mdf.py", line 1625, in filter
      cg_nr = mdf.append(
    File "C:\Python39\lib\site-packages\asammdf\blocks\mdf_v4.py", line 2777, in append
      ch = Channel(**kwargs)
    File "C:\Python39\lib\site-packages\asammdf\blocks\v4_blocks.py", line 931, in __init__
      self.attachments = 1
  AttributeError: 'Channel' object has no attribute 'attachments'
```
I believe this might be a Typo as _attachment\_nr_ is declared and never filled but _attachments_ is not declared.

Replacing line 931 with
`                self.attachment_nr = 1
`
seems to work for me, but I am not sure how this would change other use cases. Also the original channel has _attachment_nr = 0_ but is I write line 931 as _self.attachment\_nr = 0_ then it fails during the saving process as the pack misses one item.
I am a bit lost. Can someone provide insight?

Thanks,

**Reply by mini2WP** [reply]
Hi all,
I cannot help but notice that all other questions were answered very quickly, yet nothing here after almost a month.
- Is it a stupid question?
- Is it a known limitation that I should have found in the documentation?
- Is everyone just too busy at the moment?
This topic has been busting my head and I am wondering if the wall is just not breakable at this stage :)

**Reply by danielhrisca** [reply]
Hello @mini2WP , this just slipped from my focus. There was a typo indeed, `attachments_nr` was the correct attribute name. This is now fixed in the development branch and will be included in release 6.3.0
